File: src/main/java/guru/xgm/jass/formatting/block/JassGlobalsBlock.java
package guru.xgm.jass.formatting.block;

import com.intellij.formatting.*;
import com.intellij.lang.ASTNode;
import com.intellij.psi.codeStyle.CodeStyleSettings;
import org.jetbrains.annotations.NotNull;

import static com.intellij.psi.formatter.FormatterUtil.isOneOf;
import static guru.xgm.jass.psi.JassTypes.*;

public class JassGlobalsBlock extends JassBlock {
    public JassGlobalsBlock(ASTNode myNode, CodeStyleSettings codeStyleSettings) {
        super(myNode, Wrap.createWrap(WrapType.NONE, false), null, Indent.getNoneIndent(), codeStyleSettings);
    }

    @Override
    public Block makeSubBlock(@NotNull ASTNode childNode) {
        Indent indent = Indent.getNormalIndent();

        if (isOneOf(childNode, GLOBAL_VAR_DECL)) return new JassGlobalVarBlock(childNode, myCodeStyleSettings);
        if (isOneOf(childNode, GLOBALS, ENDGLOBALS)) indent = Indent.getNoneIndent();

        return new JassBlock(childNode, null, null, indent, myCodeStyleSettings);
    }

    @Override
    public @NotNull ChildAttributes getChildAttributes(int i) {
        return new ChildAttributes(Indent.getNormalIndent(), null);
    }

    @Override
    public boolean isIncomplete() {
        return !isOneOf(myNode.getLastChildNode(), ENDGLOBALS);
    }
}

File: src/main/kotlin/guru/xgm/language/jass/formatting/block/JassGlobalsBlock.kt
package guru.xgm.language.jass.formatting.block

import com.intellij.formatting.Alignment
import com.intellij.formatting.Block
import com.intellij.formatting.ChildAttributes
import com.intellij.formatting.Indent
import com.intellij.lang.ASTNode
import com.intellij.psi.codeStyle.CodeStyleSettings
import com.intellij.psi.formatter.FormatterUtil
import guru.xgm.language.jass.formatting.JassCodeStyleSettings
import guru.xgm.language.jass.formatting.block.JassGlobalVarBlock.Companion.getAlignments
import guru.xgm.language.jass.psi.JassTypes

class JassGlobalsBlock(myNode: ASTNode?, code: CodeStyleSettings) :
    JassBlock(myNode!!, null, Indent.getNoneIndent(), code) {
    private val gvarAlignments: HashMap<String, Alignment>

    init {
        val jass = code.getCustomSettings(JassCodeStyleSettings::class.java)
        gvarAlignments = getAlignments(jass)
    }

    override fun makeSubBlock(childNode: ASTNode): Block {
        var indent = Indent.getNormalIndent()

        if (FormatterUtil.isOneOf(childNode, JassTypes.GVAR)) return JassGlobalVarBlock(
            childNode,
            null,
            Indent.getNormalIndent(),
            myCodeStyleSettings,
            gvarAlignments
        )
        if (FormatterUtil.isOneOf(childNode, JassTypes.GLOBALS, JassTypes.ENDGLOBALS)) indent = Indent.getNoneIndent()

        return JassBlock(childNode, null, indent, myCodeStyleSettings)
    }

    override fun getChildAttributes(i: Int): ChildAttributes {
        return ChildAttributes(Indent.getNormalIndent(), null)
    }

    override fun isIncomplete(): Boolean {
        return !FormatterUtil.isOneOf(myNode.lastChildNode, JassTypes.ENDGLOBALS)
    }
}
